client/src/main/java/io/pravega/client/batch/impl/BatchClientImpl.java

```diff
@@ -60,43 +68,98 @@ private StreamInfo getStreamInfo(Stream stream) {
-                                                                                          from.getTime()),
-                                                             RuntimeException::new);
-        SortedSet<Segment> result = new TreeSet<>();
-        result.addAll(segments.keySet());
-        result.addAll(getAndHandleExceptions(controller.getSuccessors(new StreamCut(stream, segments)),
-                                             RuntimeException::new));
-        return Iterators.transform(result.iterator(), s -> segmentToInfo(s));
+    @Override
+    public StreamSegmentInfo getSegments(final Stream stream, final StreamCut fromStreamCut, final StreamCut toStreamCut) {
+        Preconditions.checkNotNull(stream, "stream");
+        return listSegments(stream, Optional.ofNullable(fromStreamCut), Optional.ofNullable(toStreamCut));
     }
 
-    private SegmentInfo segmentToInfo(Segment s) {
-        @Cleanup
-        SegmentMetadataClient client = segmentMetadataClientFactory.createSegmentMetadataClient(s);
-        return client.getSegmentInfo();
+    @Override
+    public <T> SegmentIterator<T> readSegment(final SegmentInputSplit segment, final Serializer<T> deserializer) {
+        return new SegmentIteratorImpl<>(inputStreamFactory, segment.getSegment(), deserializer,
+                segment.getStartOffset(), segment.getEndOffset());
     }
 
-    @Override
-    public <T> SegmentIterator<T> readSegment(Segment segment, Serializer<T> deserializer) {
-        @Cleanup
-        SegmentMetadataClient metadataClient = segmentMetadataClientFactory.createSegmentMetadataClient(segment);
-        SegmentInfo segmentInfo = metadataClient.getSegmentInfo();
-        return new SegmentIteratorImpl<>(inputStreamFactory, segment, deserializer, segmentInfo.getStartingOffset(), segmentInfo.getWriteOffset());
+    private StreamSegmentInfo listSegments(final Stream stream, final Optional<StreamCut> startStreamCut,
+                                           final Optional<StreamCut> endStreamCut) {
+        //Validate that the stream cuts are for the requested stream.
+        startStreamCut.ifPresent(streamCut -> Preconditions.checkArgument(stream.equals(streamCut.getStream())));
+        endStreamCut.ifPresent(streamCut -> Preconditions.checkArgument(stream.equals(streamCut.getStream())));
+
+        // if startStreamCut is not provided use the streamCut at the start of the stream.
+        // if toStreamCut is not provided obtain a streamCut at the tail of the stream.
+        CompletableFuture<StreamCut> startSC = startStreamCut.isPresent() ?
+                CompletableFuture.completedFuture(startStreamCut.get()) : fetchStreamCut(stream, new Date(0L));
+        CompletableFuture<StreamCut> endSC = endStreamCut.isPresent() ?
+                CompletableFuture.completedFuture(endStreamCut.get()) : fetchTailStreamCut(stream);
+        return getStreamSegmentInfo(startSC, endSC);
```

Comment:
I am not a big fan of passing `CompletableFuture` as arguments. Could we do this using `Futures.allOfWithResults(startSC, endSC)` and then invoking `getStreamSegmentInfo` with the result (using `handle` or `whenComplete`?

This will avoid the two calls to `join` in the `getStreamSegmentInfo` below

Reply:
Done, used `CompletableFuture.thenCombine(..)` and ensured we do not pass `CompletableFuture` as arguments .